## **Part 1: Understand the Business Problem**

### **The Business Problem**

ABC Manufacturing Ltd is experiencing employee attrition (employees leaving the company).

The HR department wants answers before investing money in machine learning.

Your role is to answer questions like:

1. What does the company's workforce look like?
2. Which departments have the highest employee attrition?
3. Does age influence attrition?
4. Does monthly income affect retention?
5. Does overtime influence attrition?
6. Which job roles experience the highest turnover?
7. Which variables appear important for future predictive modelling?


### **Business Objective**

Analyze the IBM HR dataset to understand workforce characteristics and identify factors associated with employee attrition through exploratory data analysis.

## **Part 2: Data Loading & Inspection**

### **Step 1: Import the necessary libraries**

In [2]:
# import liraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### **Step 2: Load the data**

In [3]:
# Load the dataset
df = pd.read_csv(r"https://drive.google.com/uc?export=download&id=16pf1k2T4Wv6g1sPSXn5kFyhRzWrUrD0u")

In [4]:
# Verify the loaded dataset
df.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,...,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,...,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,...,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,...,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,...,4,80,1,6,3,3,2,2,2,2


In [5]:
df.tail()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
1465,36,No,Travel_Frequently,884,Research & Development,23,2,Medical,1,2061,...,3,80,1,17,3,3,5,2,0,3
1466,39,No,Travel_Rarely,613,Research & Development,6,1,Medical,1,2062,...,1,80,1,9,5,3,7,7,1,7
1467,27,No,Travel_Rarely,155,Research & Development,4,3,Life Sciences,1,2064,...,2,80,1,6,0,3,6,2,0,3
1468,49,No,Travel_Frequently,1023,Sales,2,3,Medical,1,2065,...,4,80,0,17,3,2,9,6,0,8
1469,34,No,Travel_Rarely,628,Research & Development,8,3,Medical,1,2068,...,1,80,0,6,3,4,4,3,1,2


### **Description of Variables**

The dataset contains the following key variables related to employee information and attrition:

*   **Age**: Age of the employee.
*   **Attrition**: Whether the employee left the company (Yes) or not (No). This is the target variable.
*   **BusinessTravel**: Frequency of business travel for the employee (e.g., Travel_Rarely, Travel_Frequently, Non-Travel).
*   **DailyRate**: Daily rate of pay for the employee.
*   **Department**: Department the employee works in (e.g., Sales, Research & Development, Human Resources).
*   **DistanceFromHome**: Distance from home to workplace in miles.
*   **Education**: Level of education (1 'Below College' to 5 'Doctor').
*   **EducationField**: Field of education (e.g., Life Sciences, Medical, Marketing).
*   **EmployeeCount**: Number of employees (always 1 for each row, can be dropped).
*   **EmployeeNumber**: Unique employee identification number.
*   **EnvironmentSatisfaction**: Satisfaction with the work environment (1 'Low' to 4 'Very High').
*   **Gender**: Gender of the employee (Male, Female).
*   **HourlyRate**: Hourly rate of pay.
*   **JobInvolvement**: Level of job involvement (1 'Low' to 4 'Very High').
*   **JobLevel**: Job level of the employee (1 to 5).
*   **JobRole**: Role of the employee (e.g., Sales Executive, Research Scientist).
*   **JobSatisfaction**: Satisfaction with the job (1 'Low' to 4 'Very High').
*   **MaritalStatus**: Marital status (e.g., Single, Married, Divorced).
*   **MonthlyIncome**: Monthly income of the employee.
*   **MonthlyRate**: Monthly rate of pay.
*   **NumCompaniesWorked**: Number of companies worked for prior to IBM.
*   **Over18**: Whether the employee is over 18 (all 'Y', can be dropped).
*   **OverTime**: Whether the employee works overtime (Yes, No).
*   **PercentSalaryHike**: Percentage increase in salary.
*   **PerformanceRating**: Performance rating (1 'Low' to 4 'Outstanding').
*   **RelationshipSatisfaction**: Satisfaction with relationships at work (1 'Low' to 4 'Very High').
*   **StandardHours**: Standard hours of work (always 80, can be dropped).
*   **StockOptionLevel**: Level of stock option (0 to 3).
*   **TotalWorkingYears**: Total number of years worked.
*   **TrainingTimesLastYear**: Number of training times last year.
*   **WorkLifeBalance**: Work-life balance (1 'Bad' to 4 'Best').
*   **YearsAtCompany**: Total number of years at the company.
*   **YearsInCurrentRole**: Number of years in the current role.
*   **YearsSinceLastPromotion**: Number of years since the last promotion.
*   **YearsWithCurrManager**: Number of years with current manager.

### **Step 3: Data Inspection**

#### **a) Shape**

In [ ]:
# Shape
df.shape

(1470, 35)

##### **Interpretation**

*   The dataset contains 1470 rows and 35 columns.


#### **b) Columns names**

In [6]:
# identify columns names
df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EmployeeNumber', 'EnvironmentSatisfaction', 'Gender', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'Over18', 'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

#### **c) Data types**

In [7]:
# Data types
df.dtypes

,0
Age,int64
Attrition,object
BusinessTravel,object
DailyRate,int64
Department,object
DistanceFromHome,int64
Education,int64
EducationField,object
EmployeeCount,int64
EmployeeNumber,int64


##### **Interpretation**

*   The dataset includes both numerical (int64) and categorical (object) data types.
*   Most columns are numerical, but key categorical columns like 'Attrition', 'BusinessTravel', 'Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus', 'Over18', and 'OverTime' are of object type.



#### **c) Missing values**

In [8]:
# Missing values
df.isnull().sum()

,0
Age,0
Attrition,0
BusinessTravel,0
DailyRate,0
Department,0
DistanceFromHome,0
Education,0
EducationField,0
EmployeeCount,0
EmployeeNumber,0


##### **Interpretation**

*   There are no missing values in any of the columns in the dataset.


#### **d) Duplicate values**

In [9]:
# Duplicate values
df.duplicated().sum()

np.int64(0)

##### **Interpretation**

*   There are no duplicate rows in the dataset.


#### **e) Descriptive Statistics**


In [11]:
# Numerical Analysis
df.describe()

,Age,DailyRate,DistanceFromHome,Education,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,HourlyRate,JobInvolvement,JobLevel,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
count,1470.000000,1470.000000,1470.000000,1470.000000,1470.0,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,...,1470.000000,1470.0,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000
mean,36.923810,802.485714,9.192517,2.912925,1.0,1024.865306,2.721769,65.891156,2.729932,2.063946,...,2.712245,80.0,0.793878,11.279592,2.799320,2.761224,7.008163,4.229252,2.187755,4.123129
std,9.135373,403.509100,8.106864,1.024165,0.0,602.024335,1.093082,20.329428,0.711561,1.106940,...,1.081209,0.0,0.852077,7.780782,1.289271,0.706476,6.126525,3.623137,3.222430,3.568136
min,18.000000,102.000000,1.000000,1.000000,1.0,1.000000,1.000000,30.000000,1.000000,1.000000,...,1.000000,80.0,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000
25%,30.000000,465.000000,2.000000,2.000000,1.0,491.250000,2.000000,48.000000,2.000000,1.000000,...,2.000000,80.0,0.000000,6.000000,2.000000,2.000000,3.000000,2.000000,0.000000,2.000000
50%,36.000000,802.000000,7.000000,3.000000,1.0,1020.500000,3.000000,66.000000,3.000000,2.000000,...,3.000000,80.0,1.000000,10.000000,3.000000,3.000000,5.000000,3.000000,1.000000,3.000000
75%,43.000000,1157.000000,14.000000,4.000000,1.0,1555.750000,4.000000,83.750000,3.000000,3.000000,...,4.000000,80.0,1.000000,15.000000,3.000000,3.000000,9.000000,7.000000,3.000000,7.000000
max,60.000000,1499.000000,29.000000,5.000000,1.0,2068.000000,4.000000,100.000000,4.000000,5.000000,...,4.000000,80.0,3.000000,40.000000,6.000000,4.000000,40.000000,18.000000,15.000000,17.000000


##### **Interpretation**

*   **Age:** The average employee age is around 37 years, with a range from 18 to 60. The median (50th percentile) is 36, indicating a relatively balanced distribution, but the mean being slightly higher suggests a slight right skew or a few older employees. The interquartile range (IQR) from 30 to 43 shows that the middle 50% of employees are in their early 30s to early 40s.
*   **DailyRate:** Daily rates vary widely, from 102 to 102 to 1499, with an average of 802. The median is also 802. The median is also 802, suggesting a symmetric distribution for this variable.
*   **DistanceFromHome:** The average distance is about 9 miles, ranging from 1 to 29 miles. The distribution shows that 75% of employees live within 14 miles of work.
*   **EmployeeCount, Over18, StandardHours:** These columns have a count of 1470 and a standard deviation of 0, meaning all values are identical (1, 'Y', and 80 respectively). As noted previously, these do not provide any variance and can be considered for removal in modeling.
*   **EmployeeNumber:** This is a unique identifier, with 1470 unique values, as expected.
*   **JobLevel, TotalWorkingYears, YearsAtCompany, YearsInCurrentRole, YearsSinceLastPromotion, YearsWithCurrManager: **These variables describe tenure and experience. Their distributions show varying levels of experience, with many employees having relatively fewer years in these categories (e.g., median YearsAtCompany is 5, median YearsWithCurrManager is 3).
*   **MonthlyIncome:** The monthly income ranges from $1009 to 1009 to 19999, with a mean of around 6503 and a median of 6503 and a median of 4919. The mean being higher than the median indicates a right-skewed distribution, meaning there are more employees earning lower incomes and a few employees with significantly higher incomes.
*   **Satisfaction Metrics (EnvironmentSatisfaction, JobInvolvement, JobSatisfaction, PerformanceRating, RelationshipSatisfaction, WorkLifeBalance):** These are on a 1-4 or 1-5 scale (or similar). Most of their means are around 2.5 to 3, suggesting a generally moderate to high level of satisfaction across the workforce, though specific distributions would require further visualization.



In [12]:
# Categorical Analysis
df.describe(include='object')

,Attrition,BusinessTravel,Department,EducationField,Gender,JobRole,MaritalStatus,Over18,OverTime
count,1470,1470,1470,1470,1470,1470,1470,1470,1470
unique,2,3,3,6,2,9,3,1,2
top,No,Travel_Rarely,Research & Development,Life Sciences,Male,Sales Executive,Married,Y,No
freq,1233,1043,961,606,882,326,673,1470,1054


##### **Interpretation**

##### **Categorical Features (df.describe(include='object') output):**

*   **Attrition:** Has 2 unique values ('No', 'Yes'). 'No' is themost frequent (top) with 1233 occurrences (freq), confirming the class imbalance previously observed.
*   **BusinessTravel:** Has 3 unique values. 'Travel_Rarely' is the most common with 1043 occurrences
*   **Department:** Has 3 unique values. 'Research & Development' is the largest department with 961 employees.
*   **EducationField:** Has 6 unique values. 'Life Sciences' is the most prevalent field of education with 606 employees.
*   **Gender:** Has 2 unique values. 'Male' is the more frequent gender with 882 employees.
*   **JobRole:** Has 9 unique values. 'Sales Executive' is the most common job role with 326 employees.
*   **MaritalStatus:** Has 3 unique values. 'Married' is the most common marital status with 673 employees.
*   **OverTime:** Has 2 unique values. 'No' (no overtime) is more frequent with 1054 employees.









#### **f) Unique values**

In [13]:
# Unique values
df.nunique()

,0
Age,43
Attrition,2
BusinessTravel,3
DailyRate,886
Department,3
DistanceFromHome,29
Education,5
EducationField,6
EmployeeCount,1
EmployeeNumber,1470


##### **Interpretation**

*   EmployeeCount, Over18, and StandardHours each have only one unique value, suggesting they are constant and might be dropped if they don't provide any useful information.

*   EmployeeNumber has 1470 unique values, indicating it's a unique identifier for each employee.

*   Columns like Age, DailyRate, MonthlyIncome, MonthlyRate, TotalWorkingYears have a wide range of unique values, as expected for numerical features.

*   Categorical columns like Department, JobRole, EducationField, BusinessTravel, MaritalStatus have a small number of unique values, representing distinct categories.



In [18]:
import os

# Create the directory if it doesn't exist
os.makedirs('../data', exist_ok=True)

In [19]:
df.to_csv("../data/employee_attrition_cleaned.csv", index=False)

In [20]:
from google.colab import files

files.download('../data/employee_attrition_cleaned.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>